mono/mini/jit-icalls.c

```diff
@@ -96,11 +96,18 @@ mono_ldvirtfn_gshared (MonoObject *obj, MonoMethod *method)
 void
 mono_helper_stelem_ref_check (MonoArray *array, MonoObject *val)
 {
+	MonoError error;
+	gboolean isinst;
+
 	if (!array) {
 		mono_set_pending_exception (mono_get_exception_null_reference ());
 		return;
 	}
-	if (val && !mono_object_isinst (val, array->obj.vtable->klass->element_class)) {
+
+	isinst = mono_object_isinst_checked (val, array->obj.vtable->klass->element_class, &error) != NULL;
+	mono_error_raise_exception (&error); /* FIXME don't raise here */
```

Comment:
Can't use mono_raise_exception here as this is an icall that was converted to use mono_set_pending_exception.